[inductor] Optimize welford reduction

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120330

This does two things,
1) Short circuit `welford_reduce` on the first iteration to ignore the accumulator (big win for small `rnumel`)
2) Replace division with multiplication by reciprocal

Currently this is not enough to match two pass reduction with bfloat16 but it is still a significant improvement.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

Differential Revision: [D56222224](https://our.internmc.facebook.com/intern/diff/D56222224)

peterbell10 committed:
[inductor] Optimize welford reduction

This does two things,
1) On the first iteration we short circuit `welford_reduce` which speeds
2) Replace division with multiplication by reciprocal

Currently this is not enough to match two pass reduction with bfloat16
but it is still a significant improvement.

[ghstack-poisoned]

peterbell10 commented:
Unfortunately this still has worse performance than two pass reduction for many of the cases in #120184 but the regression is at least less severe. Here is the result for `bfloat16` (note fewer points are worse than eager compared to the issue)
![welford_compare](https://github.com/pytorch/pytorch/assets/13238737/237892f6-bbc3-4bc2-b52f-f8f58b7099da)

And for `float32` we have a few regressions around 3000-4000 elements
![welford_compare_float32](https://github.com/pytorch/pytorch/assets/13238737/e5b09f0f-0ee0-4871-91b8-33fd105997bb)

peterbell10 commented:
@pytorchbot merge -r

peterbell10 committed:
Update on "[inductor] Optimize welford reduction"


This does two things,
1) Short circuit `welford_reduce` on the first iteration to ignore the accumulator (big win for small `rnumel`)
2) Replace division with multiplication by reciprocal

Currently this is not enough to match two pass reduction with bfloat16 but it is still a significant improvement.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

jeffdaily commented:
@pytorchbot revert

jeffdaily commented:
@pytorchbot revert -m "broke ROCm CI while ROCm was in unstable status" -c ignoredsignal

peterbell10 committed:
Update on "[inductor] Optimize welford reduction"


This does two things,
1) Short circuit `welford_reduce` on the first iteration to ignore the accumulator (big win for small `rnumel`)
2) Replace division with multiplication by reciprocal

Currently this is not enough to match two pass reduction with bfloat16 but it is still a significant improvement.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

peterbell10 committed:
disable first-iteration opt. on rocm on "[inductor] Optimize welford reduction"


This does two things,
1) Short circuit `welford_reduce` on the first iteration to ignore the accumulator (big win for small `rnumel`)
2) Replace division with multiplication by reciprocal

Currently this is not enough to match two pass reduction with bfloat16 but it is still a significant improvement.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

lezcano commented on torch/_inductor/triton_helpers.py:
```diff
@@ -92,22 +93,42 @@ def max_with_index(value, index, dim):
-        new_weight,
-    )
+def div_approx(a, b):
+    return a * tl.math.rcp_rn(b)
```
`rcp.rn` is just a different rounding. If you want to use the fast one you should probably use `approx`. Otherwise, is `a * (1/b)` really faster than `a/b`? I've been looking around and it seems that most implementations of `a/b` do it via multiplication + inversion.

peterbell10 replied:
> rcp.rn is just a different rounding. If you want to use the fast one you should probably use approx.

`div.approx` isn't directly exposed in the triton language, however if you look at the ptx manual it states
> `div.approx.f32` implements a fast approximation to divide, computed as d = a * (1/b).
so this is equivalent.

There is also an `rcp.approx` which might speed things up further but would need inline assembly to use it.

I couldn't find latency tables for a current gen GPU but this paper gives ~2x latency improvement on a TITAN RTX. (Table IV at the bottom)
https://arxiv.org/pdf/1905.08778.pdf

peterbell10 replied:
I found a source for A100, that claims `div.rn.f32` is 525 cycles/inst vs 198 for `rcp.rn.f32`.

https://arxiv.org/pdf/2208.11174.pdf (Table V on page 7)

lezcano replied:
That table is great! Look at that `rcp.approx.f32` at 23 cycles. I wonder whether that'd be usable at all...

peterbell10 committed:
Update on "[inductor] Optimize welford reduction"


This does two things,
1) Short circuit `welford_reduce` on the first iteration to ignore the accumulator (big win for small `rnumel`)
2) Replace division with multiplication by reciprocal

Currently this is not enough to match two pass reduction with bfloat16 but it is still a significant improvement.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

kit1980 commented:
@pytorchbot revert -m "Broke internal tests, see D54230858" -c ghfist

kit1980 commented:
@pytorchbot revert -m "Broke internal tests, see D54230858" -c ghfirst

kit1980 commented:
@peterbell10 Sorry for the absence of the revert details, it's not clear what exactly happened in the moment.
The failed test is in the internal user code, not in the PyTorch itself.

kit1980 commented:
@peterbell10 
I've got suggested this hypothesis for the internal failure:

It is possible that the diff introduced rounding and precision problem here
`w2_over_w = tl.where(new_weight == 0.0, 0.0, div_approx(weight_2, new_weight))`

peterbell10 commented:
@kit1980 is it a small bump in test tolerance, or a catastrophic failure in accuracy? A minified program with inputs would be really useful here, otherwise I can't say much.

kit1980 commented:
@peterbell10 I'll find someone to follow-up.

peterbell10 committed:
Update on "[inductor] Optimize welford reduction"


This does two things,
1) Short circuit `welford_reduce` on the first iteration to ignore the accumulator (big win for small `rnumel`)
2) Replace division with multiplication by reciprocal

Currently this is not enough to match two pass reduction with bfloat16 but it is still a significant improvement.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Optimize welford reduction"


This does two things,
1) Short circuit `welford_reduce` on the first iteration to ignore the accumulator (big win for small `rnumel`)
2) Replace division with multiplication by reciprocal

Currently this is not enough to match two pass reduction with bfloat16 but it is still a significant improvement.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

Chillee committed:
Update on "[inductor] Optimize welford reduction"


This does two things,
1) Short circuit `welford_reduce` on the first iteration to ignore the accumulator (big win for small `rnumel`)
2) Replace division with multiplication by reciprocal

Currently this is not enough to match two pass reduction with bfloat16 but it is still a significant improvement.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

Chillee committed:
Update on "[inductor] Optimize welford reduction"


This does two things,
1) Short circuit `welford_reduce` on the first iteration to ignore the accumulator (big win for small `rnumel`)
2) Replace division with multiplication by reciprocal

Currently this is not enough to match two pass reduction with bfloat16 but it is still a significant improvement.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

Differential Revision: [D56222224](https://our.internmc.facebook.com/intern/diff/D56222224)

[ghstack-poisoned]

Chillee committed:
Update on "[inductor] Optimize welford reduction"


This does two things,
1) Short circuit `welford_reduce` on the first iteration to ignore the accumulator (big win for small `rnumel`)
2) Replace division with multiplication by reciprocal

Currently this is not enough to match two pass reduction with bfloat16 but it is still a significant improvement.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

Differential Revision: [D56222224](https://our.internmc.facebook.com/intern/diff/D56222224)

[ghstack-poisoned]

Chillee commented:
@chillee has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56222224).

peterbell10 commented:
After rerunning the benchmarks, the perf gain seems to have disappeared, or perhaps was previously dominated by the removal of the first iteration reduction step which was spun out into a separate PR. 

Looking at the SASS for the version using division, it seems to be doing:
```
rcp = reciprocal(b)
tmp = a * rcp
error = fma(-tmp, b, a)
result = fma(error, rcp, tmp)
```
which has only a two cycle latency increase compared to `result = a * reciprocal(b)`, so only a marginal benefit.  There is an additional out of line slow path that's used when the divisor is outside a certain range, but I don't think we hit that case here.
